basic_block.c

```diff
@@ -56,23 +43,15 @@ static const guint8 afl_maybe_log_code[] = {
  *     #include <stdint.h>
  *     #include <stdlib.h>
  *
- *     #define MAP_SIZE_POW2 16
- *     #define MAP_SIZE (1 << MAP_SIZE_POW2)
- *
  *     uint8_t** afl_area_ptr_ptr;
- *     unsigned int afl_instr_rms;
  *     uint64_t* __afl_prev_loc;
  *
- *     void afl_maybe_log(size_t current_pc) {
- *         size_t offset = (current_pc >> 4) ^ (current_pc << 8);
- *         offset &= MAP_SIZE - 1;
- *         if (offset >= afl_instr_rms) return;
- *
+ *     void afl_log(size_t offset) {
```

Comment:
you renamed afl_maybe_log to afl_log but I do not see where you change the call for non-x64?
also by `(*afl_area_ptr_ptr)[offset ^ *__afl_prev_loc]++;` you can be out of the map size? by default its 64kb but this is configurable, that is what `offset &= MAP_SIZE - 1;` is for - ensuring that the block id fits into the map.

Reply:
@vanhauser-thc This is only pseudo-code where I removed the `maybe` part to make it clear that this slimmed down version of `afl_maybe_log()` always logs, because the offset computation and check is done at the basic block's compile-time. Because `MAP_SIZE` is constant, we can evaluate that only once. So we do the same calculation ahead of time, and bake the computed offset into the basic block. We could actually do the same for non-x64, passing the offset to the callout. (Though that's out of scope for this PR.)

Reply:
> also by (*afl_area_ptr_ptr)[offset ^ *__afl_prev_loc]++; you can be out of the map size

Ooh this is a great catch, thanks!! This is a bug in the C implementation as well.

Reply:
Actually, scratch that: `__afl_prev_loc` is set to `offset >> 2` – and `offset` has already been validated, so this will never write out of bounds.